Bearded.Utilities/Algorithms/CoffmanGraham.cs

```diff
@@ -0,0 +1,190 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bearded.Utilities.Graphs;
+using Bearded.Utilities.Linq;
+
+namespace Bearded.Utilities.Algorithms
+{
+    /// <summary>
+    /// This class contains logic to take a partially ordered set of elements and split it in a sequence of layers such
+    /// that the following holds:
+    ///
+    /// <list type="number">
+    /// <item>
+    /// <description>
+    /// If an element is smaller than another in the partial ordering, it will be on a lower level.
+    /// </description>
+    /// </item>
+    /// <item>
+    /// Every layer contains at most a fixed number of W elements.
+    /// </item>
+    /// </list>
+    ///
+    /// <para>
+    /// The input is given as a directed acyclic graph, such that there is an arrow from x to y if x is smaller than y.
+    /// </para>
+    ///
+    /// <para>
+    /// The algorithm itself takes O(n^2), but requires an order that is transitively reduced. The runtime for the
+    /// transitive reduction of the directed acyclic graph is not included in the runtime, and is not known to be
+    /// possible in O(n^2).
+    /// </para>
+    /// </summary>
+    public static class CoffmanGraham
+    {
+        public class Instance<T> where T : IEquatable<T>
+        {
+            private readonly IDirectedAcyclicGraph<T> graph;
+            private readonly int maxLayerSize;
+            private readonly bool isGraphReduced;
+
+            internal Instance(IDirectedAcyclicGraph<T> graph, int maxLayerSize, bool isGraphReduced)
```

Comment:
Yes, I believe so. This class currently represents a single instance of the problem, and shouldn't be reused. It's maybe not the best design, but so far we haven't really had a satisfying design for algorithms like these I feel.

Since the `graph` and `isGraphReduced` are tightly coupled together, we would have to move both of them to the `Solve` method. At that point, I think we need to rename this class to be a `Solver`. I would in that case also suggest that we switch to polymorphism instead of a boolean parameter by creating a solver interface, and have a class that wraps the main solver that does a transitive reduction first.

Rough outline how that would look (names temporary):

```cs
class CoffmanGraham {
  interface ISolver {
    ImmutableList<ImmutableHashSet<T>> Solve<T>(IDirectedAcyclicGraph<T> graph);
  }

  private class ActualSolver : ISolver {
    // implementation here
  }

  private class WrappedSolver : ISolver {
    private readonly ActualSolver solver;

    public WrappedSolver(int maxLayerSize) {
      solver = new ActualSolver(maxLayerSize);
    }

    public ImmutableList<ImmutableHashSet<T>> Solve<T>(IDirectedAcyclicGraph<T> graph) {
      var reducedGraph = DirectedAcyclicGraphTransitiveReducer<T>.ReduceGraph(graph);
      return solver.Solve(reducedGraph);
    }
  }

  public static ISolver SolverForArbitraryGraphs(int maxLayerSize) => new WrappedSolver(maxLayerSize);
  public static ISolver SolverForAlreadyReducedGraphs(int maxLayerSize) => new ActualSolver(maxLayerSize);
}
```

The alternative is that we don't even expose the `Instance`, and just create `SolveForInstance` methods instead. Thoughts?

Reply:
I would prefer the (I)Solver approach since you could reuse that for some sort of composited algorithm.
(interface, sealed class, whatever - though the polymorphism are maybe nice because of composition, instead of branching?)

Otherwise, the public interface already looks pretty much like that anyways I suppose. You're exposing one type and two static construction methods. That's all good.

Reply:
Using `ISolver` as discussed.